### DISCUSSION TITLE
Is it possible to "close" an entire file/package?

### DISCUSSION THREAD

**Discussion by uhthomas** [cuelang/cue#857] [General]
Given a file `test.cue`
```yaml
close({})

a: "a"
```

Initially I would have expected the `close` to apply a constraint to the entire file where no fields can be set, but this does not appear to be the case.

This also means the following CUE snippet is valid
```yaml
import "k8s.io/api/core/v1"

v1.#List & {}

some_field: "some value"
```

Should it be?

**Comment by myitcv** [comment]
@uhthomas:

> Initially I would have expected the `close` to apply a constraint to the entire file where no fields can be set, but this does not appear to be the case.

I see what you're intending here. This does not work, but if it did work it would be at the package level:

https://tip.cuelang.org/docs/concepts/packages/

Files are only relevant in the context of scoping: https://tip.cuelang.org/docs/references/spec/#blocks

> This also means the following CUE snippet is valid
> 
> ```yaml
> import "k8s.io/api/core/v1"
> 
> v1.#List & {}
> 
> some_field: "some value"
> ```

Here `v1.#List & {}` is being [embedded](https://tip.cuelang.org/docs/references/spec/#embedding). The first paragraph of that section is relevant here:

> A struct may contain an _embedded value_, an operand used as a declaration. An embedded value of type struct is unified with the struct in which it is embedded, but disregarding the restrictions imposed by closed structs. So if an embedding resolves to a closed struct, the corresponding enclosing struct will also be closed, but may have fields that are not allowed if normal rules for closed structs were observed.

So the embedding does not constrain the package value in the way you are expecting, because the constraints are relaxed for the purposes of embedding, and the result (the package value in this case) then closed.

What you are looking for here is the [`must()`](https://github.com/cuelang/cue/issues/575) builtin, which would be used as follows:

```
must(v1.#List)
```

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/857.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
